### PR TITLE
For slice loops

### PR DESCRIPTION
**Pull request by frroossst**
The syntactically sugary for slice loops.

Just need your opinion on the right way of constructing the *ast.ForSlice node in forSlice()

Right now, I just blindly take whatever expression the parser gives me, should I be checking it's validation and if not valid throw a syntax error

### CODE REVIEW

File: compile/function.go [hunk 1]
```diff
@@ -360,7 +404,7 @@ func (p *Parser) forClassic() *ast.For {
 	inc := p.optExprList(tok.RParen)
 	p.Match(tok.RParen)
 	body := p.statement()
-	return &ast.For{Init: init, Cond: cond, Inc: inc, Body: body}
+	return &ast.For{Slice: false, Init: init, Cond: cond, Inc: inc, Body: body}
```
Comment (apmckinlay) [inline compile/function.go:407]: `Slice: false` is unnecessary since Go always initializes to the zero value, which for bool is false

**Comment by apmckinlay** [comment]
FYI - recently there's been a proposal for Go to add: `for i := range <int-expr>`

https://github.com/golang/go/issues/61405

**Comment by apmckinlay** [comment]
In general, lookahead should be avoided. Lookahead in a language/grammar tends to make it harder to read/understand. Admittedly, the parser already uses some lookahead, but in this case I think it's unnecessary. Lexer lookahead is also insufficient in this case because we don't want to match just `for id in number`, we want to more generally match `for id in <expression>`. That can be handled by backtracking, i.e. trying to match one construct and if it fails, backing up and trying to match the other. But backtracking can be tricky and is inefficient. And again, I don't think it's necessary in this case.

When two constructs start with the same prefix, in this case `for id in <expr>` it's better to match the prefix first, and then determine which one. If it's the new construct, that prefix will be followed by RangeTo.

File: compile/function.go [hunk 2]
```diff
@@ -348,6 +365,33 @@ func (p *Parser) forIn() *ast.ForIn {
 	return &ast.ForIn{Var: ast.Ident{Name: id, Pos: pos}, E: expr, Body: body}
 }
 
+func (p *Parser) forSlice() *ast.For {
```
Comment (apmckinlay) [inline compile/function.go:368]: "slice" makes sense in Go, but if we implement it in jSuneido that name won't make sense. Nor will it mean anything to Suneido programmers. I'd suggest something like `forCounted`

Comment (frroossst) [reply compile/function.go:368]: *a slice of the number line?* :joy: 

but no, it's minor, will rename.

**Comment by frroossst** [comment]
Hey sorry for the radio silence, first year engineering is really kicking my butt. 

So, the code from `forStmt()` wouldn't check if it is a for-counted/slice or for-in loop, it would go into `forIn()` and partway through check if it has the `tok.RangeTo` operator, and if it does, returns ForCounted/Slice struct, else it'd go ahead and return the ForIn struct. 

```
func (p *Parser) forStmt() ast.Statement {
	// easier to check before matching For so everything is ahead
	forIn := p.isForIn()
	// forSlice := p.isForSlice()
	forSlice := false
	p.Match(tok.For)
	if forSlice {
		return p.forSlice()
	} else if forIn {
		return p.forIn()
	}
	return p.forClassic()
}


func (p *Parser) forIn() *ast.ForIn {
	parens := p.MatchIf(tok.LParen)
	id := p.Text
	p.final[id] = disqualified
	pos := p.Pos
	p.MatchIdent()
	p.Match(tok.In)
	expr := p.exprExpecting(!parens)
	if parens {
		p.Match(tok.RParen)
	}
	body := p.statement()

// fmt.Println("forIn", id, expr, body)
	if p.MatchIf(tok.RangeTo) {
		return nil
		// inc := []ast.Expr{&ast.Unary{Tok: tok.Inc, E: &iter_var}}
		// return &ast.For{ Slice: true, Init: init_ast, Cond: cond_ast, Inc: inc, Body: body }
	}

	return &ast.ForIn{Var: ast.Ident{Name: id, Pos: pos}, E: expr, Body: body}
}

```
I tried to do the above but you can't have multiple return types in Go, there's some debug code in there. The above should eliminate the lookAheads in the code as the matching `for x in` is taken care by the `forIn()` func and then a branch happens.

**Comment by apmckinlay** [comment]
No worries, understandable.
I'm on a holiday road trip right now, but I'll take a look when I get a
chance.

**Pull request closed without merging**